### ISSUE TITLE
OpenSSL 1.1.0 read pipeline logic regressed constant-time CBC unpad code

### ISSUE THREAD

**Issue by davidben** [openssl/openssl#1438]
It seems the master branch may have broke the Lucky 13 fix. In 94777c9c86a2b2ea2726c49d6c8f61078558beba, the code around `tls1_cbc_remove_padding` was changed to:

```
            for (ctr = 0; ctr < numpipes; ctr++) {
                tmpret = tls1_cbc_remove_padding(s, &recs[ctr], bs, mac_size);
                if (tmpret == -1)
                    return -1;
                ret &= tmpret;
            }
```

However, `tls1_cbc_remove_padding` is documented to return:

```
 *   0: (in non-constant time) if the record is publicly invalid.
 *   1: if the padding was valid
 *  -1: otherwise.
```

0 is the return value for publicly invalid (and that `ssl3_cbc_digest_record`'s preconditions are not satisfied and must not be called). 1 and -1 must be handled in timing-independent code, not 1 and 0.

I don't understand the pipelining code, but there might even be a crash here. `ssl3_cbc_copy_mac` and `ssl3_cbc_digest_record` rely on `tls_cbc_remove_padding` to ensure some preconditions are met. (The preconditions fail when the record is publicly invalid, so it works out.) But since `tmpret` is short-circuited, if a later record is publicly invalid, the caller won't notice.

By the way, this return convention is kind of confusing, so I just made this change in BoringSSL to make it a bit clearer:
https://boringssl.googlesource.com/boringssl/+/3f26a49eb6de454cd81ae5e58e621084cd7d645c%5E%21/
(Note that patch won't be directly applicable. Our code here's diverged significantly. Notably, in OpenSSL, this also interacts with `tls1_enc`'s return value, while we do the decrypt and MAC check in one operation. We've also hidden all the legacy ciphers behind fake "AEAD"s which is why it's in libcrypto.)

**Comment by davidben** [comment]
CC @agl to make sure I didn't say anything false.

**Comment by mattcaswell** [comment]
Hmmm. Good catch. Perhaps it should look something like this:
````
            for (ctr = 0; ctr < numpipes; ctr++) {
                tmpret = tls1_cbc_remove_padding(s, &recs[ctr], bs, mac_size);
                if (tmpret == 0)
                    return 0;
                ret = constant_time_select(constant_time_lt(tmpret, 0), -1, ret);
            }
````

**Comment by davidben** [comment]
I think that works?

You might also want to consider importing, with appropriate modifications, the change to the return value here:
https://boringssl.googlesource.com/boringssl/+/3f26a49eb6de454cd81ae5e58e621084cd7d645c%5E%21/

Then you can write:
```
    unsigned padding_ok;
    if (!tls1_cbc_remove_padding(s, &padding_ok, ....)) {
        /* Publicly invalid. */
        return 0;
    }
   ret = constant_time_select(padding_ok, ret, -1);
```
which is I think a little bit less obtuse. You still have the awkward magic -1 return value, but that can be cleaned up later. (I would recommend that, post-1.1.0, you all investigate merging `tls1_enc` and `tls1_mac` into one hook. The split between `tls1_cbc_remove_padding` and `ssl3_cbc_copy_mac` is pretty weird.)

**Comment by mattcaswell** [comment]
The return value change is a bit more invasive in OpenSSL due the split between `tls1_enc` and the MAC value check. I am reluctant to make that change this close to 1.1.0 release, but perhaps for 1.1.0-next we should look at it. I will fix the immediate issue for now, but leave this issue open for review after 1.1.0 release.

**Comment by davidben** [comment]
Oh? Sorry, I should have been less lazy about the ...s and made my snippet clearer. I was just thinking something like this, which doesn't affect `tls1_enc`'s return value:

```
            for (ctr = 0; ctr < numpipes; ctr++) {
                unsigned padding_ok;
                if (!tls1_cbc_remove_padding(s, &padding_ok, &recs[ctr], bs, mac_size)) {
                    /* Publicly invalid padding. */
                    return 0;
                }
                ret = constant_time_select(padding_ok, ret, -1);
            }
```

Anyway, whatever works. Your snippet does the same thing and is certainly a more minimal fix which is possibly better this late in the release cycle.

**Comment by mattcaswell** [comment]
This is fixed in 	f9cf774, which is the minimal fix. Changing the milestone to post 1.1.0 to deal with the larger refactoring.

**Comment by levitte** [comment]
Ping @mattcaswell ?

**Comment by mattcaswell** [comment]
Still should be kept open. I am cycle starved due to the TLS1.3 work, so if someone else wants to pick it up I'd be very happy. Otherwise this will have to wait.

**Comment by t8m** [comment]
The regression was fixed in 1.1.0 already.

@mattcaswell is the refactoring mentioned above still relevant?

**Comment by mattcaswell** [comment]
This has been significantly changed in 3.0. E.g. we now have `tls1_cbc_remove_padding_and_mac` and the padding and mac removal code has been moved provider side.

I think this can be closed.
